rclcpp/include/rclcpp/detail/rmw_implementation_specific_publisher_payload.hpp

```diff
@@ -0,0 +1,48 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__DETAIL__RMW_IMPLEMENTATION_SPECIFIC_PUBLISHER_PAYLOAD_HPP_
+#define RCLCPP__DETAIL__RMW_IMPLEMENTATION_SPECIFIC_PUBLISHER_PAYLOAD_HPP_
+
+#include "rcl/publisher.h"
+
+#include "rclcpp/detail/rmw_implementation_specific_payload.hpp"
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+namespace detail
+{
```

Comment:
If someone wants to pass an RMW specific publisher/subscription payload, they will need to define a subclass of `RMWImplementationSpecificPublisherPayload` or `RMWImplementationSpecificSubscriptionPayload`, and use it in the publisher/subscription option.
So, I don't think this should be under `detail` namespace, because it's actually part of the `rclcpp` API.

If we want to be able to modify this without further notice, we should put it under `experimental`, or modify `detail` readme to be clearer:
https://github.com/ros2/rclcpp/blob/41a7e783ecbd9af41a3382576ca1229f14323a20/rclcpp/include/rclcpp/detail/README.md

Reply:
My reason for placing it outside of the public API is that it is not rmw portable and therefore not recommended.

I don't see a reason for an `experimental` namespace, it's not experimental, it's just not stable or portable or recommended.

People not using this type can just pass `nullptr` or not set it at all.

Reply:
> My reason for placing it outside of the public API is that it is not rmw portable and therefore not recommended.

I agree in that we should avoid using it, except in extreme use cases.
But still, if someone else has a custom `rmw` implementation and for some reason they also want to influence publisher/subscription construction (in a way not supported by ROS), they would also use this same class as an user for passing an specific payload.
IMO, this is part of the API and not a detail. I understand a detail as something it's used internally by the implementation, and not by an user.

If we are putting it behind `detail` namespace just because we think it's something it shouldn't be used, we shouldn't add this at all.

Reply:
> If we are putting it behind detail namespace just because we think it's something it shouldn't be used, we shouldn't add this at all.

I don't understand that logic, for me detail is simply not being subject to API stability (tick-tock between versions and stuff like that).

But if everyone wants to make it part of the public API I can move it...

Reply:
> I understand a detail as something it's used internally by the implementation, and not by an user.

Also, a user would never use this class directly, instead the rmw implementation would create a derived class of this type, the user would instantiate that, and then pass it via the payload in the options. For me this is why it's a detail, the user should never create one of these classes directly.

Reply:
> Also, a user would never use this class directly, instead the rmw implementation would create a derived class of this type

Yes, that's true. I still see any usage outside `rclcpp` as an user, but I don't mind much if this comment is addressed or not.